language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs

```diff
@@ -46,46 +54,87 @@ impl NativeVector {
-        // TODO: bounds check + implement retrieving reference to element of vector here
-        let vector_element = Value::struct_(Struct::new(vec![]));
-        let return_values = vec![vector_element];
-        NativeReturnStatus::Success {
-            cost,
-            return_values,
+    pub fn native_push_back(mut args: VecDeque<Value>) -> NativeReturnStatus {
+        if args.len() != 2 {
+            return NativeReturnStatus::InvalidArguments;
         }
+        let reference = get_vector_ref!(args);
+        let elem = match args.pop_front() {
+            Some(v) => MutVal::new(v),
+            None => return NativeReturnStatus::InvalidArguments,
+        };
+        reference
+            .mutate_native_struct(|native_val| Some(get_mut_vector(native_val)?.0.push(elem)))
+            .map(|_| NativeReturnStatus::Success {
+                cost: PUSH_BACK_COST,
+                return_values: vec![],
+            })
+            .unwrap_or(NativeReturnStatus::InvalidArguments)
     }
 
-    #[allow(unreachable_code)]
-    pub fn native_push_back(_arguments: VecDeque<Value>) -> NativeReturnStatus {
-        unimplemented!("Adding an element to a vector");
-        let cost = PUSH_BACK_COST;
-        let return_values = vec![];
-        NativeReturnStatus::Success {
-            cost,
-            return_values,
+    pub fn native_borrow(mut args: VecDeque<Value>) -> NativeReturnStatus {
+        if args.len() != 2 {
+            return NativeReturnStatus::InvalidArguments;
+        }
+        let reference = get_vector_ref!(args);
+        let idx = pop_arg!(args, u64);
+        match reference.get_native_struct_reference(|native_val| {
+            get_vector(native_val)?
+                .0
+                .get(idx as usize)
+                .map(MutVal::clone)
+        }) {
+            Some(v) => NativeReturnStatus::Success {
+                cost: BORROW_COST,
+                return_values: vec![v],
+            },
+            None => NativeReturnStatus::Aborted {
+                cost: BORROW_COST,
+                error_code: INDEX_OUT_OF_BOUND,
+            },
+        }
+    }
+
+    pub fn native_pop(mut args: VecDeque<Value>) -> NativeReturnStatus {
+        if args.len() != 1 {
+            return NativeReturnStatus::InvalidArguments;
+        }
+
+        let reference = get_vector_ref!(args);
+        match reference.mutate_native_struct(|native_val| {
+            get_mut_vector(native_val)?.0.pop().map(MutVal::into_value)
+        }) {
+            // Vector is already empty.
+            None => NativeReturnStatus::Aborted {
+                cost: POP_COST,
+                error_code: POP_EMPTY_VEC,
+            },
+            Some(Ok(v)) => {
+                println!("Popped {:?}", v);
+                NativeReturnStatus::Success {
+                    cost: POP_COST,
+                    return_values: vec![v],
+                }
+            }
+            // The poped element has dangling references.
```

Comment:
```suggestion
            // The popped element has dangling references.
```

Reply:
When can this happen? Should it be prevented by the bytecode verifier?

Reply:
Exactly. This should be prevented by the verifier.

Reply:
Got it, thanks! Should we add a special `InternalError` error code here like we do for the other errors that we expect to be impossible due to the verifier?

Reply:
Discussed offline; `InvalidArguments` seems to be a catchall type for internal errors that should be prevented by the bytecode verifier.